src/time/Waveform.cpp

```diff
@@ -20,110 +22,159 @@ int Waveform::getInterpolationOrder() const
-  _numberOfStoredSamples = 1; // the first sample is automatically initialized as zero and stored.
-  _storageIsInitialized  = true;
-  PRECICE_ASSERT(this->maxNumberOfStoredSamples() == storageSize);
-  PRECICE_ASSERT(this->valuesSize() == values.size());
-  for (int sampleIndex = 0; sampleIndex < maxNumberOfStoredSamples(); ++sampleIndex) {
-    this->storeAt(values, sampleIndex);
+  _timeStepsStorage[0.0] = Eigen::VectorXd(values);
+  _timeStepsStorage[1.0] = Eigen::VectorXd(values);
+}
+
+void Waveform::store(const Eigen::VectorXd &values, double normalizedDt)
+{
+  PRECICE_ASSERT(_timeStepsStorage.size() > 0);
+  // dt has to be in interval (0.0, 1.0]
+  PRECICE_ASSERT(normalizedDt > 0.0); // cannot override value at beginning of window. It is locked!
+  PRECICE_ASSERT(normalizedDt <= 1.0);
+
+  if (math::equals(maxStoredDt(), 1.0)) {          // reached end of window and trying to write new data from next window. Clearing window first.
+    Eigen::VectorXd keep = _timeStepsStorage[0.0]; // we keep data at _timeStepsStorage[0.0]
+    _timeStepsStorage.clear();
+    _timeStepsStorage[0.0] = keep;
+  } else { // did not reach end of window yet, so dt has to strictly increase
+    PRECICE_ASSERT(normalizedDt > maxStoredDt(), normalizedDt, maxStoredDt());
   }
+  PRECICE_ASSERT(values.size() == _timeStepsStorage[0.0].size());
+  this->_timeStepsStorage[normalizedDt] = Eigen::VectorXd(values);
 }
 
-void Waveform::store(const Eigen::VectorXd &values)
+double Waveform::maxStoredDt()
 {
-  PRECICE_ASSERT(_storageIsInitialized);
-  int sampleIndex = 0;
-  this->storeAt(values, sampleIndex);
+  double maxDt = -1;
+  for (auto timeStep : _timeStepsStorage) {
+    if (timeStep.first > maxDt) {
+      maxDt = timeStep.first;
+    }
+  }
+  PRECICE_ASSERT(maxDt >= 0);
+  return maxDt;
 }
 
-void Waveform::storeAt(const Eigen::VectorXd values, int sampleIndex)
+// helper function to compute x(t) from given data (x0,t0), (x1,t1), ..., (xn,tn) via B-spline interpolation (implemented using Eigen).
+Eigen::VectorXd bSplineInterpolationAt(double t, Eigen::VectorXd ts, Eigen::MatrixXd xs, int splineDegree)
 {
-  PRECICE_ASSERT(_timeWindowsStorage.cols() > sampleIndex, maxNumberOfStoredSamples(), sampleIndex);
-  PRECICE_ASSERT(values.size() == this->valuesSize(), values.size(), this->valuesSize());
-  this->_timeWindowsStorage.col(sampleIndex) = values;
+  // organize data in columns. Each column represents one sample in time.
+  PRECICE_ASSERT(xs.cols() == ts.size());
+  const int ndofs = xs.rows(); // number of dofs. Each dof needs it's own interpolant.
+
+  Eigen::VectorXd interpolated(ndofs);
+
+  const int splineDimension = 1;
+
+  for (int i = 0; i < ndofs; i++) {
+    auto spline     = Eigen::SplineFitting<Eigen::Spline<double, splineDimension>>::Interpolate(xs.row(i), splineDegree, ts);
+    interpolated[i] = spline(t)[0]; // get component of spline associated with xs.row(i)
+  }
```

Comment:
I did not check, but isn't there a vectorized function for this in Eigen? (such that we don't need the loop)

Reply:
Not sure, but I already spent so much time on Eigen here that I don't want to spend more time on runtime optimizations. Error messages are really difficult to understand in Eigen. I would rather put this into an issue and see whether we really need it. (If somebody else wants to dig deeper here, I'm also fine.)

Reply:
I think I remember the actual reason now: `Eigen::SplineFitting<Eigen::Spline<double, splineDimension>>` is a template. This means: `splineDimension` has to be known at compile time.